agent/taskresource/credentialspec/credentialspec_windows.go

```diff
@@ -0,0 +1,609 @@
+	return s, nil
+}
+
+func (cs *CredentialSpecResource) initStatusToTransition() {
+	resourceStatusToTransitionFunction := map[resourcestatus.ResourceStatus]func() error{
+		resourcestatus.ResourceStatus(CredentialSpecCreated): cs.Create,
+	}
+	cs.resourceStatusToTransitionFunction = resourceStatusToTransitionFunction
+}
+
+func (cs *CredentialSpecResource) Initialize(resourceFields *taskresource.ResourceFields,
+	taskKnownStatus status.TaskStatus,
+	taskDesiredStatus status.TaskStatus) {
+	cs.initStatusToTransition()
+	cs.credentialsManager = resourceFields.CredentialsManager
+	cs.ssmClientCreator = resourceFields.SSMClientCreator
+	cs.s3ClientCreator = resourceFields.S3ClientCreator
+
+	// if task hasn't turn to 'created' status, and it's desire status is 'running'
+	// the resource status needs to be reset to 'NONE' status so the cs value
+	// will be retrieved again
+	if taskKnownStatus < status.TaskCreated &&
+		taskDesiredStatus <= status.TaskRunning {
+		cs.SetKnownStatus(resourcestatus.ResourceStatusNone)
+	}
+}
+
+// GetTerminalReason returns an error string to propagate up through to task
+// state change messages
+func (cs *CredentialSpecResource) GetTerminalReason() string {
+	return cs.terminalReason
+}
+
+func (cs *CredentialSpecResource) setTerminalReason(reason string) {
+	cs.terminalReasonOnce.Do(func() {
+		seelog.Infof("credentialspec resource: setting terminal reason for credentialspec resource in task: [%s]", cs.taskARN)
+		cs.terminalReason = reason
+	})
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (cs *CredentialSpecResource) GetDesiredStatus() resourcestatus.ResourceStatus {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.desiredStatusUnsafe
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (cs *CredentialSpecResource) SetDesiredStatus(status resourcestatus.ResourceStatus) {
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	cs.desiredStatusUnsafe = status
+}
+
+// DesiredTerminal returns true if the credentialspec's desired status is REMOVED
+func (cs *CredentialSpecResource) DesiredTerminal() bool {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.desiredStatusUnsafe == resourcestatus.ResourceStatus(CredentialSpecRemoved)
+}
+
+// KnownCreated returns true if the credentialspec's known status is CREATED
+func (cs *CredentialSpecResource) KnownCreated() bool {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.knownStatusUnsafe == resourcestatus.ResourceStatus(CredentialSpecCreated)
+}
+
+// TerminalStatus returns the last transition state of credentialspec
+func (cs *CredentialSpecResource) TerminalStatus() resourcestatus.ResourceStatus {
+	return resourcestatus.ResourceStatus(CredentialSpecRemoved)
+}
+
+// NextKnownState returns the state that the resource should
+// progress to based on its `KnownState`.
+func (cs *CredentialSpecResource) NextKnownState() resourcestatus.ResourceStatus {
+	return cs.GetKnownStatus() + 1
```

Comment:
Can we not use magic number?

Reply:
This is part of the same boilerplate used in multiple places within the repo
Refer:
* https://github.com/aws/amazon-ecs-agent/blob/master/agent/taskresource/firelens/firelens_unix.go#L323
* https://github.com/aws/amazon-ecs-agent/blob/master/agent/taskresource/ssmsecret/ssmsecret.go#L162

Reply:
This is not a big concern for me for now. I will resolve it.